pkg/util/cryptomaterial/certchains/signers.go

```diff
@@ -14,287 +14,297 @@ import (
+		return fmt.Errorf("failed to regenerate cert %q: %v", certInfo.GetMeta().Name, err)
+	}
 
-type CertificateSigningRequestInfo struct {
-	Name         string
-	ValidityDays int
+	return nil
 }
 
-type signedCertificateInfo struct {
-	certDir   string
-	tlsConfig *crypto.TLSCertificateConfig
-}
+func (s *CertificateSigner) AddToBundles(bundlePaths ...string) error {
+	cert := s.signerConfig.Config.Certs[0]
 
-// NewCertificateSigner returns a builder object for a certificate chain for the given signer
-func NewCertificateSigner(signerName, signerDir string, validityDays int) *certificateSigner {
-	return &certificateSigner{
-		signerName:         signerName,
-		signerDir:          signerDir,
-		signerValidityDays: validityDays,
-	}
-}
+	for _, bundlePath := range bundlePaths {
+		bundlePEMs, err := os.ReadFile(bundlePath)
+		if err != nil && !os.IsNotExist(err) {
+			return err
+		}
 
-func (s *certificateSigner) WithClientCertificates(signInfos ...*ClientCertificateSigningRequestInfo) *certificateSigner {
-	s.clientCertificatesToSign = append(s.clientCertificatesToSign, signInfos...)
-	return s
-}
+		certs := []*x509.Certificate{}
+		if len(bundlePEMs) > 0 {
+			certs, err = crypto.CertsFromPEM(bundlePEMs)
+			if err != nil {
+				return err
+			}
+		}
 
-func (s *certificateSigner) WithServingCertificates(signInfos ...*ServingCertificateSigningRequestInfo) *certificateSigner {
-	s.serverCertificatesToSign = append(s.serverCertificatesToSign, signInfos...)
-	return s
-}
+		var certsChanged, certFound bool
+		for i, c := range certs {
+			if c.Subject.String() == cert.Subject.String() &&
+				c.Issuer.String() == cert.Issuer.String() {
+
+				certFound = true
+				if c.SerialNumber != cert.SerialNumber {
+					certs[i] = cert
+					certsChanged = true
+				}
+				break
+			}
+		}
 
-func (s *certificateSigner) WithPeerCertificiates(signInfos ...*PeerCertificateSigningRequestInfo) *certificateSigner {
-	s.peerCertificatesToSign = append(s.peerCertificatesToSign, signInfos...)
-	return s
-}
+		if certFound {
+			if !certsChanged {
+				continue
+			}
+		} else {
+			certs = append(certs, cert)
+		}
 
-func (s *certificateSigner) WithSubCAs(subCAsInfo ...*certificateSigner) *certificateSigner {
-	s.subCAs = append(s.subCAs, subCAsInfo...)
-	return s
-}
+		// make sure the parent directory exists
+		if err := os.MkdirAll(filepath.Dir(bundlePath), os.FileMode(0755)); err != nil {
+			return err
+		}
 
-func (s *certificateSigner) Complete() (*CertificateSigner, error) {
-	// in case this is a sub-ca, it's already going to have the signer-config populated
-	signerConfig := s.signerConfig
-	if signerConfig == nil {
-		var err error
-		signerConfig, _, err = crypto.EnsureCA(
-			CACertPath(s.signerDir),
-			CAKeyPath(s.signerDir),
-			CASerialsPath(s.signerDir),
-			s.signerName,
-			s.signerValidityDays,
-		)
+		certFileWriter, err := os.OpenFile(bundlePath, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0644)
```

Comment:
@oglok this part should hopefully fix the ever-growing CA bundle files

Reply:
nice! 👍